core/templates/pages/exploration-player-page/services/question-player-engine.service.spec.ts

```diff
@@ -0,0 +1,815 @@
+
+    questionPlayerEngineService.clearQuestions();
+    totalQuestions = questionPlayerEngineService.getQuestionCount();
+    expect(totalQuestions).toBe(0);
+
+    questionPlayerEngineService.init(
+      [singleQuestionBackendDict], successHandler, failHandler);
+    totalQuestions = questionPlayerEngineService.getQuestionCount();
+    expect(totalQuestions).toBe(1);
+  });
+
+  it('should clear all questions', () => {
+    let successHandler = jasmine.createSpy('success');
+    let failHandler = jasmine.createSpy('fail');
+
+    spyOn(contextService, 'setQuestionPlayerIsOpen').and.returnValue(null);
+    spyOn(contextService, 'isInQuestionPlayerMode').and.returnValue(true);
+
+    questionPlayerEngineService.init(
+      multipleQuestionsBackendDict, successHandler, failHandler);
+
+    expect(questionPlayerEngineService.getQuestionCount()).toBe(3);
+
+    questionPlayerEngineService.clearQuestions();
+
+    expect(questionPlayerEngineService.getQuestionCount()).toBe(0);
+  });
+
+  it('should return the language code correctly when new card is added', () => {
+    let successCallback = jasmine.createSpy('success');
+    let successHandler = jasmine.createSpy('success');
+    let failHandler = jasmine.createSpy('fail');
+    let answer = 'answer';
+    let interactionRulesService: InteractionRulesService;
+    let answerClassificationResult = new AnswerClassificationResult(
+      outcomeObjectFactory
+        .createNew('default', '', '', []), 1, 0, 'default_outcome'
+    );
+
+    spyOn(contextService, 'setQuestionPlayerIsOpen').and.returnValue(null);
+    spyOn(contextService, 'isInQuestionPlayerMode').and.returnValue(true);
+    spyOn(answerClassificationService, 'getMatchingClassificationResult')
+      .and.returnValue(answerClassificationResult);
+    spyOn(expressionInterpolationService, 'processHtml')
+      .and.callFake((html, envs) => html);
+
+    questionPlayerEngineService.init(
+      multipleQuestionsBackendDict, successHandler, failHandler);
+    let languageCode = questionPlayerEngineService.getLanguageCode();
+
+    expect(languageCode).toBe('en');
+
+    questionPlayerEngineService.submitAnswer(
+      answer, interactionRulesService, successCallback);
+    questionPlayerEngineService.recordNewCardAdded();
+
+    languageCode = questionPlayerEngineService.getLanguageCode();
+    expect(languageCode).toBe('br');
+
+    questionPlayerEngineService.submitAnswer(
+      answer, interactionRulesService, successCallback);
+    questionPlayerEngineService.recordNewCardAdded();
+
+    languageCode = questionPlayerEngineService.getLanguageCode();
+    expect(languageCode).toBe('ab');
+  });
+
+  it('should always return false when calling \'isInPreviewMode()\'', () => {
+    let previewMode = questionPlayerEngineService.isInPreviewMode();
+
+    expect(previewMode).toBe(false);
+  });
+
+  it('should show warning message while loading a question ' +
+    'if the question name is empty', () => {
```

Comment:
Is this a duplicate test (see L683)?

Reply:
Nope,
this check happens when loading a question
another one will happen after submitting a question
(Please note the describe block `describe('on submitting answer '`)

Reply:
I see, thanks!